src/tainting/Taint.ml

```diff
@@ -94,30 +75,129 @@ let _show_arg { pos = s, i; offset = os } =
     Printf.sprintf "arg(%s)#%d.%s" s i os_str
   else Printf.sprintf "arg(%s)#%d" s i
 
+(*****************************************************************************)
+(* Preconditions *)
+(*****************************************************************************)
+
+(* We roll our own boolean formula type here because we need to be
+   able to use polymorphic compare. Comparison on the Generic AST
+   isn't possible, and polymorphic compare would otherwise take into account
+   unimportant details like tokens.
+*)
+type precondition =
+  | Label of string
+  | Bool of bool
+  | And of precondition list
+  | Or of precondition list
+  | Not of precondition
+[@@deriving show]
+
+let rec expr_to_precondition e =
+  match e.G.e with
+  | G.L (G.Bool (v, _)) -> Bool v
+  | G.N (G.Id ((str, _), _)) -> Label str
+  | G.Call ({ e = G.IdSpecial (G.Op G.Not, _); _ }, (_, [ Arg e1 ], _)) ->
+      Not (expr_to_precondition e1)
+  | G.Call ({ e = G.IdSpecial (G.Op op, _); _ }, (_, args, _)) -> (
+      match (op, args_to_precondition args) with
+      | G.And, xs -> And xs
+      | G.Or, xs -> Or xs
+      | __else__ ->
+          logger#error "Unexpected Boolean operator";
+          Bool false)
+  | G.ParenExpr (_, e, _) -> expr_to_precondition e
+  | ___else__ ->
+      logger#error "Unexpected `requires' expression";
+      Bool false
+
+and args_to_precondition args =
+  match args with
+  | [] -> []
+  | G.Arg e :: args' -> expr_to_precondition e :: args_to_precondition args'
+  | _ :: args' ->
+      logger#error "Unexpected argument kind";
+      Bool false :: args_to_precondition args'
+
 (*****************************************************************************)
 (* Taint *)
 (*****************************************************************************)
 
-type orig = Src of source | Arg of arg [@@deriving show]
-type taint = { orig : orig; tokens : tainted_tokens } [@@deriving show]
+type source = {
+  call_trace : Rule.taint_source call_trace;
+  label : string;
+      (* This is needed because we may change the label of a taint,
+         from the original source that it came from.
+         This happens from propagators which change the label of the taint.
+         We don't put it under `taint`, though, because Arg taints are
+         supposed to be polymorphic in label.
+      *)
+  precondition : (taint list * precondition) option;
+}
+[@@deriving show]
 
-let src_of_pm (pm, (x : Rule.taint_source)) =
-  Src { call_trace = PM (pm, x); label = x.label }
+and orig = Src of source | Arg of arg [@@deriving show]
+and taint = { orig : orig; tokens : tainted_tokens } [@@deriving show]
+
+let substitute_precondition_arg_taint ~arg_fn taint =
+  let rec aux taint =
+    match taint.orig with
+    | Arg arg -> arg_fn arg
+    | Src ({ precondition = None; _ } as src) ->
+        [ { taint with orig = Src src } ]
```

Comment:
Isn't this the same as just `[ taint ]` ?

Reply:
yep, you're correct. dunno how that happened.